File: backend/app/schemas/workspaces.py
# =============================================================================
# File: workspaces.py
# Module/Service: Workspace Service
# Layer: Schema
# Purpose: Pydantic request/response models for Workspace CRUD (FR1 / OpenAPI).
# Responsibilities:
#   - Match OpenAPI Workspace, WorkspaceListResponse, create/update bodies
# Dependencies:
#   - Pydantic
# Public Exports:
#   - WorkspaceCreateRequest, WorkspaceUpdateRequest
#   - WorkspaceResponse, WorkspaceListResponse
# Database/Table: workspaces
# Related Modules: app.api.workspaces, docs/Enterprise_notebooklm_openapi.yaml
# Important Notes: deleted_at is internal — not exposed in API response schema.
# =============================================================================

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class WorkspaceCreateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=2000)


class WorkspaceUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=2000)


class WorkspaceResponse(BaseModel):
    id: UUID
    name: str
    description: str | None = None
    created_at: datetime
    updated_at: datetime


class WorkspaceListResponse(BaseModel):
    items: list[WorkspaceResponse]
    page: int
    page_size: int
    total: int

File: frontend/types/workspaces.ts
/**
 * =============================================================================
 * File: workspaces.ts
 * Module/Service: Workspace Service (Web App)
 * Layer: UI
 * Purpose: TypeScript types for Workspace CRUD matching OpenAPI (FR1).
 * Responsibilities:
 *   - Align frontend Workspace models with backend Pydantic schemas
 * Dependencies:
 *   - docs/Enterprise_notebooklm_openapi.yaml
 * Public Exports:
 *   - Workspace, WorkspaceListResponse, WorkspaceCreateInput, WorkspaceUpdateInput
 *   - WorkspaceMember, MemberCandidate, AddMemberInput, UpdateMemberRoleInput
 * Database/Table: N/A
 * Related Modules: lib/api-client, features/workspaces
 * Important Notes: deleted_at is not exposed in API responses.
 * =============================================================================
 */

import type { WorkspaceRole } from "./auth";

export type Workspace = {
  id: string;
  name: string;
  description: string | null;
  created_at: string;
  updated_at: string;
};

export type WorkspaceListResponse = {
  items: Workspace[];
  page: number;
  page_size: number;
  total: number;
};

export type WorkspaceCreateInput = {
  name: string;
  description?: string | null;
};

export type WorkspaceUpdateInput = {
  name?: string;
  description?: string | null;
};

/** Matches OpenAPI WorkspaceMember (GET/POST/PATCH .../members). */
export type WorkspaceMember = {
  user_id: string;
  email: string;
  role: WorkspaceRole;
  joined_at: string;
};

/** Active user eligible to invite (GET .../member-candidates). */
export type MemberCandidate = {
  user_id: string;
  email: string;
  full_name: string;
};

/**
 * POST /workspaces/{id}/members — provide user_id and/or email
 * (backend requires at least one).
 */
export type AddMemberInput = {
  user_id?: string;
  email?: string;
  role: WorkspaceRole;
};

export type UpdateMemberRoleInput = {
  role: WorkspaceRole;
};
